pyo3-macros-backend/src/method.rs

```diff
@@ -125,8 +125,14 @@ impl FnType {
             FnType::FnClass(span) | FnType::FnNewClass(span) => {
                 let py = syn::Ident::new("py", Span::call_site());
                 let slf: Ident = syn::Ident::new("_slf", Span::call_site());
+                let allow_deprecated = if cfg!(feature = "gil-refs") {
+                    quote!()
+                } else {
+                    quote!(#[allow(deprecated)])
+                };
                 quote_spanned! { *span =>
                     #[allow(clippy::useless_conversion)]
+                    #allow_deprecated
```

Comment:
I am not sure if I follow things correctly: If the feature is enabled, the GIL ref methods are not deprecated and we do need to suppress the warning. If the feature is not disabled, then we do suppress it. But I thought we actually want to either warn about the usage of the deprecated API or use the new API in that case? Is it impossible to avoid the GIL ref API here?

Reply:
I stuck with the GIL-Ref API here because I feel like the unconstrained `'a` lifetime in `from_type_ptr_borrowed` is waiting to introduce accidental unsoundness, and I didn't feel comfortable putting it in the macro output here. I want to come up with a better way to solve this one than straight-up switch to `from_type_ptr_borrowed` (and I'm also ok if we block merging this PR until we figure out a solution).

Reply:
The simplest I can think of right now is to put a `where 'py: 'a` bound on `from_type_ptr_borrowed`, which at least works for this case and helps add a guard rail against arbitrarily long "borrowed references" on the type ptr. I'd still love to have something better if possible.

Reply:
I am wondering again if should have this bound on the `Borrowed` type itself just as we it is implicitly present for `&'a T`, `Borrowed<'a, 'py, T>` standing in for `&'a Bound<'py, T>` should have a `Bound<'py, T>: 'a` bound, meaning at least `'py: 'a`.

Of course, we do need to extend the lifetime at some place, but I suspect having to use unsafe there might be preferable to having this foot gun present in an `Borrowed`-based API.

Reply:
Yes, this might be a tidy solution for now. It's probably non breaking to relax that constraint again later.

I'll try to push a commit later today which implements this adjustment so we can see how it looks.

Reply:
Pushed as a third commit; I definitely don't love the constraint, I had to add 4 transmutes to extend the lifetimes back for the `Py<PyString>` and `Py<PyBytes>` access.

I think the main thing which might inform a better design is what type would we prefer users to use for `#[classmethod]`? It's a borrowed reference, so should they take `cls: Borrowed<PyType>`, or `cls: &Bound<PyType>`?

For now as implemented they would take `Borrowed<PyType>` but `&Bound<PyType>` might be more consistent with what we're doing elsewhere, and the reference would necessarily be bounded by the duration the pointer is on the stack, which would provide an alternative route to making this safe.

Reply:
> Pushed as a third commit; I definitely don't love the constraint, I had to add 4 transmutes to extend the lifetimes back for the Py<PyString> and Py<PyBytes> access.

I agree that the `transmute`s are bad, but we could also add a more misuse resistant helper method which would only ever extend the first lifetime. That said...

> I think the main thing which might inform a better design is what type would we prefer users to use for #[classmethod]? It's a borrowed reference, so should they take cls: Borrowed<PyType>, or cls: &Bound<PyType>?

> For now as implemented they would take Borrowed<PyType> but &Bound<PyType> might be more consistent with what we're doing elsewhere, and the reference would necessarily be bounded by the duration the pointer is on the stack, which would provide an alternative route to making this safe.

...I agree that using `&Bound<'py, PyType>` might be the preferable approach here. The main issue I see is that we need to somehow document and enforce our stance towards `Borrowed`, e.g. "it should never be used in a public API if `&Bound` works as well" or "used only to prevent otherwise necessary but generally superfluous reference increments". Could we maybe even make `Borrowed` unnameable for downstream crates so that it can only be accessed as a return value or via type interference on trait impls?

Reply:
> Could we maybe even make `Borrowed` unnameable for downstream crates so that it can only be accessed as a return value or via type interference on trait impls?

I'm certainly up for trying that, and even if we decide to make `Borrowed` more public later, it does help keep down the number of new things we're introducing to users in a single release.

I'll push one or two more commits later tonight to switch to `&Bound` here, and take `Borrowed` out of reachable paths for now. Let's see how that works out.

Reply:
I haven't forgotten about this; just had illness and subsequent exhaustion in the family. Might be a couple of days before I have energy to get to this still 😭

Reply:
Finally completed the `&Bound<PyType>` bit in #3831, so I'll rebase here.